### PR TITLE
Add OpenVINO IR inference for CLS, OD, SSEG, IS tasks + OMZ OV models validation support  

### PR DESCRIPTION
**Pull request by kprokofi**
### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/engine/test.py [hunk 1]
```diff
@@ -33,7 +34,7 @@ def test(cfg: TrainConfig) -> tuple[Trainer, dict[str, Any]]:
     trainer: Trainer = hydra.utils.instantiate(cfg.trainer)
 
     log.info("Starting testing!")
-    if cfg.checkpoint is None:
+    if not isinstance(model.model.model, Model) and cfg.checkpoint is None:
```
Comment (vinnamkim) [inline src/otx/core/engine/test.py]: Why do we need this change?

File: src/otx/core/data/entity/base.py [hunk 2]
```diff
@@ -157,10 +157,6 @@ def collate_fn(cls, entities: list[T_OTXDataEntity]) -> OTXBatchDataEntity:
             msg = "collate_fn() input should include a single OTX task"
             raise RuntimeError(msg)
 
-        if not all(entity.image_type == ImageType.TV_IMAGE for entity in entities):
```
Comment (vinnamkim) [inline src/otx/core/data/entity/base.py:175]: Why do we need this change?

File: src/otx/core/model/entity/classification.py [hunk 3]
```diff
@@ -217,3 +217,55 @@ def _customize_outputs(
...
+    def forward(
+        self,
+        inputs: MulticlassClsBatchDataEntity,
+    ) -> MulticlassClsBatchPredEntity | OTXBatchLossEntity:
+        """Model forward function."""
+        # If customize_inputs is overrided
+        if self.training:
+            msg = "OV model cannot be used for training"
+            raise RuntimeError(msg)
+        if len(inputs.images) > 1:
```
Comment (vinnamkim) [inline src/otx/core/model/entity/classification.py]:
```suggestion
        if inputs.batch_size > 1:
```

File: src/otx/core/model/entity/classification.py [hunk 4]
```diff
@@ -217,3 +217,55 @@ def _customize_outputs(
...
+
+    def _customize_outputs(
+        self,
+        outputs: Any,  # noqa: ANN401
+        inputs: MulticlassClsBatchDataEntity,
+    ) -> MulticlassClsBatchPredEntity:
+        # add label index
+        labels = [outputs.top_labels[0][0]]
+        # add probability
+        scores = [outputs.top_labels[0][2]]
```
Comment (vinnamkim) [inline src/otx/core/model/entity/classification.py]: Are these `torch.Tensor`?

File: src/otx/core/model/entity/detection.py [hunk 5]
```diff
@@ -143,3 +142,67 @@ def _customize_outputs(
...
+        return DetectionModel.create_model(self.model_name)
+
+    def _customize_outputs(
+        self,
+        outputs: Any,  # noqa: ANN401
+        inputs: DetBatchDataEntity,
+    ) -> DetBatchPredEntity | OTXBatchLossEntity:
+        # add label index
+        outputs_objects = outputs.objects
+        bboxes = [
```
Comment (vinnamkim) [inline src/otx/core/model/entity/detection.py:186]: This should be `tv_tensor.BoundingBoxes`.

File: src/otx/core/model/entity/detection.py [hunk 6]
```diff
@@ -143,3 +142,67 @@ def _customize_outputs(
...
+            torch.tensor(
+                [[output.xmin, output.ymin, output.xmax, output.ymax] for output in outputs_objects],
+                dtype=torch.float32,
+            ),
+        ]
+        scores = [torch.tensor([output.score for output in outputs_objects])]
+
+        if self.model.get_label_name(0) == "background":
+            # some OMZ models require labels shift to match target labels
```
Comment (vinnamkim) [inline src/otx/core/model/entity/detection.py]: Please add warning for this special case.

File: src/otx/core/model/module/segmentation.py [hunk 7]
```diff
@@ -31,7 +31,7 @@ def __init__(
         super().__init__(otx_model, optimizer, scheduler, torch_compile)
         metric_params = {
             "task": "multiclass",
-            "num_classes": otx_model.model.decode_head.num_classes,
+            "num_classes": otx_model.config.get("decode_head", {}).get("num_classes", None),
```
Comment (vinnamkim) [inline src/otx/core/model/module/segmentation.py]: Is this can be nullable?

https://github.com/Lightning-AI/torchmetrics/blob/3514d715ef42b8dc0c8d96cb3d581caba512fbd7/src/torchmetrics/classification/jaccard.py#L462

It seems that we cannot pass `None` for `multiclass`. Please try to fail here.

File: src/otx/recipe/detection/atss_mobilenetv2.yaml [hunk 8]
```diff
@@ -132,7 +132,7 @@ model:
         relu_before_extra_convs: true
       bbox_head:
         type: ATSSHead
-        num_classes: 2
+        num_classes: 20
```
Comment (vinnamkim) [inline src/otx/recipe/detection/atss_mobilenetv2.yaml]: Why do we need this?

File: tests/unit/core/model/module/test_segmentation.py [hunk 9]
```diff
@@ -33,9 +33,7 @@ class TestOTXSegmentationModel:
     def model(self, mocker, fxt_seg_data_entity) -> OTXSegmentationLitModule:
         # define otx model
         otx_model = mocker.MagicMock(spec=MMSegCompatibleModel)
-        otx_model.model = mocker.MagicMock(spec=torch.nn.Module)
-        otx_model.model.decode_head = mocker.MagicMock(spec=torch.nn.Module)
-        otx_model.model.decode_head.num_classes = 2
+        otx_model.config = {"decode_head": {"num_classes": 2}}
```
Comment (vinnamkim) [inline tests/unit/core/model/module/test_segmentation.py:36]: Is it necessary?

File: src/otx/core/data/transform_libs/torchvision.py [hunk 10]
```diff
@@ -9,39 +9,63 @@
...
 
             if type(transform) not in availables:
                 msg = f"transform={transform} is not a valid TorchVision V2 transform"
                 raise ValueError(msg)
 
             transforms.append(transform)
 
+        if return_as_list:
+            return transforms
+
```
Comment (vinnamkim) [inline src/otx/core/data/transform_libs/torchvision.py]: I'm against of adding if-else branches agglomeratively to the existing code. Why don't you make a new transform such as `transform = Identity()`?

File: src/otx/core/engine/test.py (hunk 1 above)
Comment (kprokofi) [reply src/otx/core/engine/test.py]: Because Openvino model runs without any checkpoints, it is a final representation of a model. So, with this condition we couldn't run "otx test" without checkpoint

File: src/otx/core/data/entity/base.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/core/data/entity/base.py:175]: Because OV IR should receive Numpy images, not torch vision tensors. This condition raised an error

File: src/otx/core/model/entity/detection.py (hunk 5 above)
Comment (kprokofi) [reply src/otx/core/model/entity/detection.py:186]: It can work perfectly without tv_tensor.BoundingBoxes. However, for the types consistency I can change it

File: src/otx/core/model/module/segmentation.py (hunk 7 above)
Comment (kprokofi) [reply src/otx/core/model/module/segmentation.py]: We can raise an exception in case of "num_classes" is None

File: tests/unit/core/model/module/test_segmentation.py (hunk 9 above)
Comment (kprokofi) [reply tests/unit/core/model/module/test_segmentation.py:36]: Yes, to pass the unit tests since I've changed the logic of how to get the number of classes.

File: src/otx/core/data/transform_libs/torchvision.py (hunk 10 above)
Comment (kprokofi) [reply src/otx/core/data/transform_libs/torchvision.py]: I've already added a new transform Identity(). However, to use it we need register it in "list_available_transforms" and then, we should return not the torchvision Compose, but a list. So, how do you propose to change "generate" method not using "if-else" statements. 
Maybe we could create separate TransformLib like TorchVisionTransformLib, but OVTransformLib. What do you think?

Comment (vinnamkim) [reply src/otx/core/data/transform_libs/torchvision.py]: > Maybe we could create separate TransformLib like TorchVisionTransformLib, but OVTransformLib. What do you think?

Yes, that looks good

File: src/otx/core/model/entity/detection.py (hunk 5 above)
Comment (vinnamkim) [reply src/otx/core/model/entity/detection.py:186]: > However, for the types consistency I can change it

Yes, please keep it consistent and care about the runtime type mismatching.

File: src/otx/core/data/entity/base.py (hunk 2 above)
Comment (vinnamkim) [reply src/otx/core/data/entity/base.py:175]: There is no `np.ndarray` type in `OTXBatchDataEntity.images`. Please make it consistent.

File: src/otx/core/engine/test.py (hunk 1 above)
Comment (vinnamkim) [reply src/otx/core/engine/test.py]: For a new comer who sees this first time, it is really hard to understand its underlying behavior. Can we enhance this code clearer? We may utilize `cfg.checkpoint` to load the OV IR file. Even if it would be not exactly this idea, I think it should be improved.

File: src/otx/core/model/module/segmentation.py (hunk 7 above)
Comment (vinnamkim) [reply src/otx/core/model/module/segmentation.py]: Please raise an exception in that case with a comprehensible message such as `msg = 'if num_classes=None, cannot get JaccardIndex("multiclass") metric'`.

File: src/otx/core/engine/test.py (hunk 1 above)
Comment (kprokofi) [reply src/otx/core/engine/test.py]: I see two options here:
1) instead of model_name we can use checkpoint to load IR model
2) remove raising error in case when checkpoint is None. In some cases even for PyTorch models we may want to use pure model with pre-trained weights to check the score on dataset

File: src/otx/core/data/entity/base.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/core/data/entity/base.py:175]: I will add np.array to OTXBatchDataEntity.images, bacause it is essential to use numpy arrays for OV model

File: src/otx/config/data/openvino.yaml [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+train_subset:
+  transform_lib_type: TORCHVISION
+  transforms:
+    - _target_: otx.core.data.transform_libs.torchvision.Identity
+  batch_size: 1
+val_subset:
+  transform_lib_type: TORCHVISION
+  transforms:
+    - _target_: otx.core.data.transform_libs.torchvision.Identity
```
Comment (sovrasov) [inline src/otx/config/data/openvino.yaml]: Why do we have to use underscores here? It looks as some kind of irregularity

Comment (kprokofi) [reply src/otx/config/data/openvino.yaml]: It comes from Hydra syntaxis. With "__ target __ " we initialize target class
https://hydra.cc/docs/advanced/instantiate_objects/overview/#instantiation-of-builtins

File: src/otx/core/data/entity/base.py (hunk 2 above)
Comment (vinnamkim) [reply src/otx/core/data/entity/base.py:175]: Can we maintain the type of `OTXBatchDataEntity.images` as `tv_tensor.Image` only?
I confirmed that we can avoid data copying costs between `np.ndarray` and `tv_tensors.Image` conversion.
Please see the following code. It shows that they share the same pointer to the internal data buffer.

```python
import numpy as np
import torch
from torchvision.tv_tensors import Image

np_tensor = np.zeros([2, 2])
tv_image = Image(data=torch.from_numpy(np_tensor))
back_to_np = tv_image.numpy()
np_tensor += 1

print("np_tensor", np_tensor)
print("tv_image", tv_image)
print("back_to_np", back_to_np)
```

```console
np_tensor [[1. 1.]
 [1. 1.]]
tv_image Image([[[1., 1.],
        [1., 1.]]], dtype=torch.float64, )
back_to_np [[[1. 1.]
  [1. 1.]]]
```

File: src/otx/core/engine/test.py (hunk 1 above)
Comment (vinnamkim) [reply src/otx/core/engine/test.py]: Let's remove this check

Comment (kprokofi) [reply src/otx/core/engine/test.py]: 1st option seems to raise problems inside lightning, OV model can't be loaded in this way

File: src/otx/core/model/module/segmentation.py (hunk 7 above)
Comment (kprokofi) [reply src/otx/core/model/module/segmentation.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2748/commits/fa36076ef72b7ad990a68b70db33ed4eb007a889)

File: src/otx/core/model/entity/detection.py (hunk 6 above)
Comment (kprokofi) [reply src/otx/core/model/entity/detection.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2748/commits/fa36076ef72b7ad990a68b70db33ed4eb007a889) 
I added it in model creation block to avoid showing warning every time during validation

File: src/otx/core/model/entity/detection.py (hunk 5 above)
Comment (kprokofi) [reply src/otx/core/model/entity/detection.py:186]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2748/commits/fa36076ef72b7ad990a68b70db33ed4eb007a889)

File: src/otx/core/model/entity/classification.py (hunk 4 above)
Comment (kprokofi) [reply src/otx/core/model/entity/classification.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2748/commits/fa36076ef72b7ad990a68b70db33ed4eb007a889)

File: src/otx/core/model/entity/classification.py (hunk 3 above)
Comment (kprokofi) [reply src/otx/core/model/entity/classification.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2748/commits/fa36076ef72b7ad990a68b70db33ed4eb007a889)

File: src/otx/core/data/entity/base.py (hunk 2 above)
Comment (kprokofi) [reply src/otx/core/data/entity/base.py:175]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2748/commits/fa36076ef72b7ad990a68b70db33ed4eb007a889) 
I used ToImage transform from torchvision. I removed Identity since it is not needed anymore. In "_customize_inputs" I added conversion from torch.tensort to np.array

File: src/otx/core/engine/test.py (hunk 1 above)
Comment (kprokofi) [reply src/otx/core/engine/test.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2748/commits/fa36076ef72b7ad990a68b70db33ed4eb007a889)

File: src/otx/core/data/transform_libs/torchvision.py (hunk 10 above)
Comment (kprokofi) [reply src/otx/core/data/transform_libs/torchvision.py]: [done](https://github.com/openvinotoolkit/training_extensions/pull/2748/commits/fa36076ef72b7ad990a68b70db33ed4eb007a889) 
It isn't required anymore

File: src/otx/core/engine/test.py [hunk 12]
```diff
@@ -33,9 +33,6 @@ def test(cfg: TrainConfig) -> tuple[Trainer, dict[str, Any]]:
     trainer: Trainer = hydra.utils.instantiate(cfg.trainer)
 
     log.info("Starting testing!")
-    if cfg.checkpoint is None:
```
Comment (sungmanc) [inline src/otx/core/engine/test.py:36]: So, in this case, users can `test` without checkpoint?

File: src/otx/core/model/entity/detection.py [hunk 13]
```diff
@@ -143,3 +144,65 @@ def _customize_outputs(
...
+    def __init__(self, config: DictConfig) -> None:
+        self.model_name = config.pop("model_name")
+        self.config = config
+        super().__init__()
+
+    def _create_model(self) -> nn.Module:
+        from openvino.model_api.models import DetectionModel
+
+        model = DetectionModel.create_model(self.model_name)
+        if model.get_label_name(0) == "background":
```
Comment (sungmanc) [inline src/otx/core/model/entity/detection.py]:
```suggestion
        if model.get_label_name(0).lower() == "background":
```
I suggest to use `lower()` function to cover the "Background" case.

**Review by vinnamkim — DISMISSED** [review]
LGTM

File: src/otx/core/engine/test.py (hunk 12 above)
Comment (kprokofi) [reply src/otx/core/engine/test.py:36]: Yes, just with pretrained weights. It is useful sometimes to check model and what the random score or just to test pre-trained weights

**Review by vinnamkim — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]

**Review by jaegukhyun — APPROVED** [review]

File: src/otx/core/model/module/segmentation.py [hunk 14]
```diff
@@ -29,9 +29,15 @@ def __init__(
         torch_compile: bool,
     ):
         super().__init__(otx_model, optimizer, scheduler, torch_compile)
+        num_classes = otx_model.config.get("decode_head", {}).get("num_classes", None)
```
Comment (harimkang) [inline src/otx/core/model/module/segmentation.py:32]: Here's a question: as far as I know, an OTXModel doesn't always hold a config, does it?

**Review by MarkByun — APPROVED** [review]

File: src/otx/core/model/module/segmentation.py (hunk 14 above)
Comment (kprokofi) [reply src/otx/core/model/module/segmentation.py:32]: Hmm, as far as I know, we always have  "config" in our recipes and rely on it heavily. What cases do you mean?
